[WIP] frontend: Email redesign.

vaidap opened the pull request:
Fixes #5113.

Current style:
![image](https://user-images.githubusercontent.com/13666710/27598498-9ced4016-5b5d-11e7-9bd7-8e60e222e3c8.png)

smarx commented:
*Automated message from Dropbox CLA bot*

@vaidap, it looks like you've already signed the Dropbox CLA. Thanks!

jrowan commented on templates/zerver/emails/email_base.html:
```diff
@@ -1,17 +1,350 @@
+        }
+
+        /* -------------------------------------
+        OTHER STYLES THAT MIGHT BE USEFUL
```
Do we need to put in all these extra styles?

vaidap replied:
i ported all of Brock's changes without refactoring-- will remove the ones that are unnecessary for now.

jrowan commented on templates/zerver/emails/email_base.html:
```diff
@@ -1,17 +1,350 @@
+                            {% block illustration %}{% endblock %}
+                        </div>
+                        <!-- START CENTERED WHITE CONTAINER -->
+                        <span class="preheader">This is preheader text. Some clients will show this text as a preview.</span>
```
I'm not sure whether my PR about preheaders will merge first and make this moot, but this should probably be a jinja2 template `{% block preheader %}{% endblock %}` so we don't default to sending emails with placeholder preheader text.

vaidap replied:
changed it to `{% block preheader %}{% endblock %}` for now!

jrowan commented on templates/zerver/emails/email_base.html:
```diff
@@ -1,17 +1,350 @@
+                                <tr>
+                                    <td class="content-block">
+                                        <span class="apple-link">Kandra Labs, 112 S Park St, San Francisco CA 94107</span>
+                                        <br/> Don't like these emails? <a href="#">Unsubscribe</a>.
```
Do we still need the unsubscribe block here if we have the `{% block manage_preferences %}` above?  Can we roll the Kandra Labs address into that area (maybe after the email-specific manage preferences content)?

vaidap replied:
one problem i've been having with both the graphics and the `manage_preferences` block is that they keep floating upwards-- see screenshots in #illustrations stream. my hacky solution right now was to remove `manage_preferences` and just keep the kandra labs footer, and insert each image manually in each email, but that is far from optimal. would it be possible for you to have a look at this & see whether you can figure out what is causing this behaviour? i'm thinking this has to do with the order in which the template blocks are generated?

vaidap replied:
it would make sense to make use of `manage_preferences` as it is adequately put in the followup emails, i completely agree. just having some trouble!

jrowan commented on templates/zerver/emails/followup_day1.html:
```diff
@@ -26,6 +26,7 @@
     {% endif %}
 
     <p>Thanks,<br/>Zulip</p>
+<img src="{{ realm_uri }}/static/generated/footer.svg"/>
```
Do we only want this footer image in this particular email, or should it be in all the emails?

vaidap commented:
not sure why `tools/lint` was failing, I had to change the CSS to this format: 
```
            img {
            border: none;
            -ms-interpolation-mode: bicubic;
            max-width: 100%;
            }
```
which is not as readable, so i added some extra spaces between the blocks. not sure if this is something that should be fixed?

vaidap commented:
Here's the current state of things! `digest` and `missed_message` haven't been modified as they do not extend the jinja2 template, and i haven't inserted the illustrations into `followup_day1` as we are planning on rewriting the text for it.
![image](https://user-images.githubusercontent.com/13666710/27823312-f91fa6b2-60a8-11e7-985d-75095d7c7099.png)
![image](https://user-images.githubusercontent.com/13666710/27823323-010eae9a-60a9-11e7-8182-ddd7813830a1.png)
![image](https://user-images.githubusercontent.com/13666710/27823329-06828392-60a9-11e7-81d1-c478167ee609.png)
![image](https://user-images.githubusercontent.com/13666710/27823357-1cf9ec3c-60a9-11e7-9aee-fb61f8767e47.png)
![image](https://user-images.githubusercontent.com/13666710/27823372-2b255224-60a9-11e7-8181-27c7c1e123b5.png)
![image](https://user-images.githubusercontent.com/13666710/27823376-30182fe0-60a9-11e7-9add-588da5521140.png)
![image](https://user-images.githubusercontent.com/13666710/27823382-35ea5e5c-60a9-11e7-98e7-96683fa39b6b.png)
![image](https://user-images.githubusercontent.com/13666710/27823390-3b9cace2-60a9-11e7-998f-8ac0a7250e6a.png)
![image](https://user-images.githubusercontent.com/13666710/27823399-41385ce6-60a9-11e7-80ac-563457bc55a1.png)

jrowan commented on templates/zerver/emails/password_reset.html:
```diff
@@ -15,3 +15,10 @@
 </p>
 {% endblock %}
+
+{% block manage_preferences %}
```
I feel like we don't want/need a manage preferences block on the password reset email since it is a one-off email specifically requested by the user.  I did a quick search through my inbox and there wasn't an unsubscribe link in any of the password reset emails I've gotten lately.

vaidap replied:
Good point, removed!

jrowan commented on templates/zerver/emails/notify_new_login.html:
```diff
@@ -15,3 +19,10 @@
 <p>Thanks,<br />Zulip Account Security</p>
 {% endblock %}
+
+{% block manage_preferences %}
```
We probably shouldn't put an unsubscribe link here since this seems like the kind of thing that should be mandatory to send/receive, or at least we probably want this to be an "unsubscribe from new login notifications" type link/messaging if we allow unsubscribing.

vaidap replied:
replaced this with a `unsubscribe from new login notifications`.

jrowan commented on templates/zerver/emails/notify_change_in_email.html:
```diff
@@ -1,12 +1,23 @@
 </p>
 {% endblock %}
+
+{% block manage_preferences %}
```
As with password reset emails, we probably don't want an unsubscribe link here because this is a one-off user-requested email related to their account security.

vaidap commented:
Regarding `{{ unsubscribe_link }}`, I'm not sure one would feed it this data?

jrowan commented on zerver/views/user_settings.py:
```diff
@@ -50,6 +50,9 @@ def confirm_email_change(request, confirmation_key):
                    'new_email': new_email,
+                    'unsub_welcome' : notifications.one_click_unsubscribe_link(user_profile,"welcome"),
+                    'unsub_digest' : notifications.one_click_unsubscribe_link(user_profile,"digest"),
+                    'unsub_missed' : notifications.one_click_unsubscribe_link(user_profile,"missed messages"),
```
We usually write this as `from zerver.lib.notifications import one_click_unsuscribe_link` at the top so we don't need to write all the `notifications.`s here (it's not super-necessary to know where the `one_click_unsubscribe_link` is coming from)

vaidap replied:
thanks for this!

vaidap commented:
regarding updating the context:
the followup emails are being fed an unsubscribe link in `notifications.py`:
```
    context.update({
        'unsubscribe_link': unsubscribe_link
    })
    send_future_email(
        "zerver/emails/followup_day1", '%s <%s>' % (name, email), from_name=from_name,
        from_address=from_address, context=context, delay=datetime.timedelta(hours=1))
```
right now, they are the only emails with unsubscribe links.

The illustrations make use of `realm_uri` and are in all the emails-- i did a check of what emails needed context updating due to this and changed the `frontend: Add header images to emails.` commit.

A few points:
* confirm_new_email has a `'realm': user_profile.realm` context assignment already; changed the image link to to `realm.uri` to match this. similarly for `notify_change_in_email`.
* `invitation_reminder` has a `context = common_context(referrer)` assignment, which already includes `realm_uri.` similarly for `followup_day1`, `followup_day2` and `notify_new_login`.

* `confirm_registration` & `invitation` amongst others don't have  a `support_email` fed into the context-- am i getting confused or should this be added?
example: 
```
    context = {'realm': user_profile.realm, 'old_email': old_email, 'new_email': new_email,
               'activate_url': activation_url}
    send_email('zerver/emails/confirm_new_email', new_email, from_name='Zulip Account Security',
               from_address=FromAddress.NOREPLY, context=context)
```

vaidap commented:
also made appropriate changes to the footers.

vaidap commented:
currently getting this error:
```
context={'activate_url': activation_url, 'realm_uri': prereg_user.realm.uri})
AttributeError: 'NoneType' object has no attribute 'uri'
```
i suppose this is because in the test, prereg_user is not set up with a realm. is there a way to get around this?

rishig commented:
cool, thanks @vaidap! I would not make any context / content changes to the emails in this PR, since it is already such a complicated PR. E.g. emails that currently don't have an unsubscribe or manage preferences link can stay without them.

Does that simplify things?

rishig commented:
Jotting down remaining work to take this over the finish line:

* Replace server_uri with email_images_base_uri (see #6262).
* Remove any context or content changes.
* Separate out the CSS into its own file, and use the premailer package to do CSS inlining (or if we think that isn't working, find a different package). Some discussion at https://chat.zulip.org/#narrow/stream/development.20help/topic/css.20inlining, but for the first pass I would just do what is easiest to code. 

Nice to do's, if they aren't too much work (but can also be done after the parts above are merged):
* Using the bubbles signoff for confirm_registration, invitation, and day1 (instead of the ~ :fish: ~ signoff) (I'll add this image to the assets directory)
* Move the jellyfish-with-a-box from day2 to day1
* Move the four fish on day2 to the top, and move the heavy fish to the bottom (i.e. to replace the ~ :fish: ~ footer)

@hackerkid, thanks!

timabbott commented:
Closing this in favor of its rebase #6383.
